storm-client/src/jvm/org/apache/storm/utils/Utils.java

```diff
@@ -1039,7 +1039,7 @@ public static Double parseJvmHeapMemByChildOpts(List<String> options, Double def
                 }
                 Matcher m = optsPattern.matcher(option);
                 while (m.find()) {
-                    int value = Integer.parseInt(m.group(1));
+                    long value = Long.parseLong(m.group(1));
```

Comment:
Even for several gigabytes, the number is still smaller than Integer.MAX_VALUE iin term of "m", "k" or "g" units, right?

Reply:
The problem is from https://github.com/apache/storm/blob/ab6b6f69a10bf57c1b46a6519d57ca423d99558c/storm-client/src/jvm/org/apache/storm/utils/Utils.java#L1105

`2 * 1024 * 1024 * 1024= 2^31 > (2^31-1) = Integer.MAX_VALUE`

Reply:
this expression "should" be promoted to double - wonder why that is not not happening.

Reply:
`value * unit` has integer overflow so it equals to `-2147483648`. 
Then `-2147483648 / 1024.0 / 1024.0 = -2048.0`, this is when it is casted to double.

Reply:
Tested this - RHS is not getting promoted to a double. But the following works with "int value:"

    double result = 1.0 * value * unit / 1024.0 / 1024.0;

       (without "1.0 *" the new case test fails)

Reply:
@ethanlm that is definitely what is happening - but I find this surprising, the expression should be widened to a double (set at compile time because of doubles in the expression (1024.0).Maybe there is a fine print at https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html.

Reply:
https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.7

> The left-hand operand of a binary operator appears to be fully evaluated before any part of the right-hand operand is evaluated.
> The Java programming language guarantees that every operand of an operator (except the conditional operators &&, ||, and ? :) appears to be fully evaluated before any part of the operation itself is performed.

So I would look at `value * unit / 1024.0 / 1024.0; ` as 
```
int tmp = value * unit;
double tmp1 = tmp / 1024.0;
double tmp2 = tmp1 / 1024.0;
```

And the Widening Primitive Conversion happens at each operation. Looking at the byte code of this short example

```

```
public static void main(String[] args)  {
        int value = 2;
        int unit = 1024 * 1024 * 1024;
        double result = value * unit / 1024.0 / 1024.0 ;
        System.out.println(result);
}
```
  public static void main(java.lang.String[]);
    Code:
       0: iconst_2
       1: istore_1
       2: ldc           #2                  // int 1073741824
       4: istore_2
       5: iload_1
       6: iload_2
       7: imul
       8: i2d
       9: ldc2_w        #3                  // double 1024.0d
      12: ddiv
      13: ldc2_w        #3                  // double 1024.0d
      16: ddiv
      17: dstore_3
      18: getstatic     #5                  // Field java/lang/System.out:Ljava/io/PrintStream;
      21: dload_3
      22: invokevirtual #6                  // Method java/io/PrintStream.println:(D)V
      25: return
}

```
It first calculate `value * unit` based on integer. Then on line8, `i2d` is invoked to convert integer to double.

Reply:
@Ethanlm thanks for the detailed bytecode explanation. This behavior is the same in jdk14 and jdk16EA. I checked via jshell.